sw/device/lib/dif/dif_otbn.c

```diff
@@ -0,0 +1,290 @@
+  register_value = bitfield_bit32_write(register_value, bit_index,
+                                        (enable == kDifOtbnEnable));
+  mmio_region_write32(otbn->base_addr, OTBN_INTR_ENABLE_REG_OFFSET,
+                      register_value);
+
+  return kDifOtbnOk;
+}
+
+dif_otbn_result_t dif_otbn_irq_force(const dif_otbn_t *otbn,
+                                     dif_otbn_interrupt_t irq_type) {
+  if (otbn == NULL) {
+    return kDifOtbnBadArg;
+  }
+
+  uint8_t bit_index;
+  if (!irq_bit_index_get(irq_type, &bit_index)) {
+    return kDifOtbnError;
+  }
+
+  // Force the requested interrupt.
+  uint32_t register_value =
+      mmio_region_read32(otbn->base_addr, OTBN_INTR_TEST_REG_OFFSET);
+  register_value = bitfield_bit32_write(register_value, bit_index, true);
+  mmio_region_write32(otbn->base_addr, OTBN_INTR_TEST_REG_OFFSET,
+                      register_value);
+
+  return kDifOtbnOk;
+}
+
+dif_otbn_result_t dif_otbn_start(const dif_otbn_t *otbn,
+                                 unsigned int start_addr) {
+  if (otbn == NULL || start_addr % 4 != 0 ||
+      start_addr >= OTBN_IMEM_SIZE_BYTES) {
+    return kDifOtbnBadArg;
+  }
+
+  mmio_region_write32(otbn->base_addr, OTBN_START_ADDR_REG_OFFSET, start_addr);
+
+  uint32_t cmd_reg_val = 0x0u;
+  cmd_reg_val = bitfield_bit32_write(cmd_reg_val, OTBN_CMD_START, true);
+  mmio_region_write32(otbn->base_addr, OTBN_CMD_REG_OFFSET, cmd_reg_val);
+
+  return kDifOtbnOk;
+}
+
+dif_otbn_result_t dif_otbn_is_busy(const dif_otbn_t *otbn, bool *busy) {
+  if (otbn == NULL || busy == NULL) {
+    return kDifOtbnBadArg;
+  }
+
+  uint32_t status = mmio_region_read32(otbn->base_addr, OTBN_STATUS_REG_OFFSET);
+  *busy =
+      bitfield_field32_read(status, (bitfield_field32_t){
+                                        .mask = 1, .index = OTBN_STATUS_BUSY,
+                                    });
+
+  return kDifOtbnOk;
+}
+
+dif_otbn_result_t dif_otbn_get_err_code(const dif_otbn_t *otbn,
+                                        dif_otbn_err_code_t *err_code) {
+  if (otbn == NULL || err_code == NULL) {
+    return kDifOtbnBadArg;
+  }
+
+  uint32_t err_code_raw =
+      mmio_region_read32(otbn->base_addr, OTBN_ERR_CODE_REG_OFFSET);
+
+  // Ensure that all values returned from hardware match explicitly defined
+  // values in the DIF.
+  switch (err_code_raw) {
+    case 0:  // kDifOtbnErrorCodeNoError
+      *err_code = (dif_otbn_err_code_t)err_code_raw;
+      return kDifOtbnOk;
+
+    default:
+      return kDifOtbnUnexpectedData;
+  }
+}
+
+dif_otbn_result_t dif_otbn_imem_write(const dif_otbn_t *otbn,
+                                      uint32_t offset_bytes,
+                                      const uint32_t *src, size_t len_bytes) {
+  // Only 32b-aligned 32b word accesses are allowed.
+  if (otbn == NULL || src == NULL || len_bytes % 4 != 0 ||
+      offset_bytes % 4 != 0 ||
+      offset_bytes + len_bytes > OTBN_IMEM_SIZE_BYTES) {
```

Comment:
I was originally writing a comment about how I didn't like `offset_bytes` and `len_bytes`, but I think I've convinced myself that callers doing:

```c
static const uint32_t kProgram[] = { ... };
const dif_otbn_result_t result = dif_otbn_imem_write(otbn, offset, kProgram, sizeof(kProgram));
```

Is probably the common pattern, and every call site doing a `sizeof(kProgram) / sizeof(uint32_t)` is more fraught than letting the `_write`/`_read` functions take care of it.

Reply:
Yeah, no fantastic solutions here. The main challenge is consistency between imem and dmem accesses, which have different word width (and hence size and alignment requirements). Having imem_write specify the size in 32b words, where dmem_write uses 256b words, feels very fragile and easy to get wrong.